### DISCUSSION TITLE
Getting Started

### DISCUSSION THREAD

**Discussion by james-machale-lseg** [OpenSourceRisk/Engine#268] [General]
A discussion for advice with regards to getting up and running with ORE.

**Comment by Alx8888** [comment]
Hi,
It looks really nice at first glance and I would like to give "the python version" a proper spin.
Are the examples mentioned in the userguide still available? I found a different set in the GitHub ORE-SWIG clone.
I assume that is the best place to start?
Many thanks!

**Reply by james-machale-lseg** [reply]
Hi! The examples from the userguide will all work with the python ORE in a notebook, but they're all designed as a push-the-button-and-it-runs examples rather than being broken down into the various steps of running ORE as the notebook examples show.

If you want to try them out in a notebook then I'd recommend starting a jupyter notebook/jupyter lab instance in the directory of e.g. example 1, and creating a little notebook. E.g. I copied this [utilities file](https://github.com/OpenSourceRisk/ORE-SWIG/blob/master/OREAnalytics-SWIG/Python/Examples/Notebooks/utilities.py) from the SWIG examples into the example 1 directory and made this tiny notebook in jupyterlab:

![image](https://github.com/user-attachments/assets/fe85e898-2680-4f56-8e54-1604ff188493)
![image](https://github.com/user-attachments/assets/62d20873-40d6-47fb-879c-34c451bad167)

Here's the code:
```python
from ORE import *
import utilities

params = Parameters()
params.fromFile("Input/ore.xml")

ore = OREApp(params)

ore.run()

display(utilities.format_report(ore.getReport('npv')))

swap_exposure = utilities.format_report(ore.getReport('exposure_trade_Swap_20y'))
display(swap_exposure)
swap_exposure.plot(x='Time', y=['EPE', 'ENE'], rot=45.0, marker='.')
```

You can definitely break up the run into smaller pieces like the notebook examples show, but I'll leave that to you!

**Comment by Alx8888** [comment]
Hi James, many thanks for your reply,

I was looking in the wrong place initially. I now found all examples mentioned in the user guide under Engine/Examples.
At first I was looking under ORE-SWIG/OREAnalytics-SWIG/Python/Examples which contains some nice notebooks, similar to what you have done above, but with different examples to ones in the userguide.

The userguide refers to that path, on page 30. "Typical usage of the Python wrapper is shown in ORE’s Example_42 and in ORE
SWIG’s OREAnalytics/Python/Examples directory"

Thanks again!

**Comment by thobitz** [comment]
Hi,

Would it be possible to provide binaries for at least the common operating systems? Or did I miss them? For someone who just wants to use the Python integration, it is a lot of effort to set up all the C++ tools.

**Reply by rkapl123** [reply]
Yes, you did miss at least the python integration, it's already on pypi: [https://pypi.org/project/open-source-risk-engine/](https://pypi.org/project/open-source-risk-engine/)

**Reply by rkapl123** [reply]
Concerning the binaries, this was discussed in #190 ad #259 that there won't be any binaries (except for the python wheels) in the github repository, however it's easy to get the binaries automatically when you fork the repo and include the github workflows for the OS you're interested in. 

E.g. for windows, take [https://github.com/OpenSourceRisk/Engine/actions/runs/19080301363/workflow](https://github.com/OpenSourceRisk/Engine/actions/runs/19080301363/workflow) and remove everything after (and including) `- name: Wrap ORE-SWIG for Python Wheels`.
Add instead following lines:
```
    - name: Save ore as artifact
      uses: actions/upload-artifact@v4
      with:
        name: ore-windows-${{ matrix.arch }}
        path: ./build/App/Release/*.*
```

After running the action (which takes a while), your binary is available from the artifacts.

**Reply by thobitz** [reply]
Thank you! I did try the pip install, but I got some error on version not satisfied, so I assumed it was missing the underlying engine - but it seems it just did not like Python 3.14. Thanks a lot for the quick help!

**Comment by thobitz** [comment]
Hi,

I noticed a few issues and/ or potential inconsistencies when trying to build under Windows as per video/ user guide. After 1 1/2 hrs a few error messages - not sure whether it worked or not. Any advice?

1) The doc says to use the "native" shell, the video the cross-tool shell - what is the right one?

2) The build breaks unless SWIG_EXECUTABLE is just "swig.exe" - it does not accept the fully qualified path 

3) I am swamped with warnings of conversions, in particular between size_t and int - is that correct?
C:\sde\ore\ore\QuantLib\ql\math\interpolations\mixedinterpolation.hpp(220,40): warning C4267: 'argument': conversion from 'size_t' to 'const int', possible loss of data [C:\sde\ore\ore\build\ORE-SWIG\OREP.vcxproj]

4) You need to install Doxygen and hence Latex and Ghostscript before, **and include the respective \bin directories in the path.** This requirement seems undocumented.

5) You **must** use Boost 1.86, not the most current 1.89

6) Some random errors and warnings - looks like a few additional settings required, like

CUSTOMBUILD : warning : Inheritance graph for 'ore::analytics::Analytic::Impl' not generated, too many nodes (27), thre
shold is 20. Consider increasing DOT_GRAPH_MAX_NODES. [C:\sde\ore\ore\build\OREAnalytics\doc\doc_orea.vcxproj]

C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): e
rror MSB8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\c8c5ed994f8075529695f9ad8f9e8f27\doc_orea.rule' exited
with code -1073741819. [C:\sde\ore\ore\build\OREAnalytics\doc\doc_orea.vcxproj]

  C:/sde/ore/ore/QuantExt/qle/instruments/forwardbond.hpp:57: warning: Found unknown command '\Name'
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): e
rror MSB8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\68fad14f002bf8ca171d655fdd364117\doc_quantext.rule' exi
ted with code -1073741819. [C:\sde\ore\ore\build\QuantExt\doc\doc_quantext.vcxproj]

And at the very end: 

"C:\sde\ore\ore\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\sde\ore\ore\build\OREAnalytics\doc\doc_orea.vcxproj" (default target) (8) ->
(CustomBuild target) ->
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5):
 error MSB8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\c8c5ed994f8075529695f9ad8f9e8f27\doc_orea.rule' exite
d with code -1073741819. [C:\sde\ore\ore\build\OREAnalytics\doc\doc_orea.vcxproj]


"C:\sde\ore\ore\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\sde\ore\ore\build\OREData\doc\doc_ored.vcxproj" (default target) (9) ->
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5):
 error MSB8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\3145e9d6192695ec9034ed807ce79c29\doc_ored.rule' exite
d with code -1073741819. [C:\sde\ore\ore\build\OREData\doc\doc_ored.vcxproj]


"C:\sde\ore\ore\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\sde\ore\ore\build\QuantExt\doc\doc_quantext.vcxproj" (default target) (10) ->
  C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5):
 error MSB8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\68fad14f002bf8ca171d655fdd364117\doc_quantext.rule' e
xited with code -1073741819. [C:\sde\ore\ore\build\QuantExt\doc\doc_quantext.vcxproj]

    556 Warning(s)
    3 Error(s)

Time Elapsed 01:22:39.63

**Reply by james-machale-lseg** [reply]
1) I'm on x64 windows and I've always used x64 native

2) this kind of issue usually very particular to how you have arranged things on your own machine!

3) yes, the conversion warnings are expected/ignored. It bothers me too, but realistically the numbers used in these types in ORE are unlikely to exceed the 32 signed integer limit.

4) Fair. I'll pass this on.

5) I'll enquire about this. We try to ensure ORE is compatible with the latest boost version but there is usually a lag as we don't have an automated system to identify breaks with the boost head.

6) we build ORE on windows every day and night and do not receive such errors. According to this [microsoft article](https://learn.microsoft.com/en-us/answers/questions/379441/error-code-exited-with-code-1073741819) , that code is for an access violation... something seems to have gone quite wrong with your local copy. Were you doing a clean set up or did you make local changes? Do you have enough HD space (>100GB for a debug build, if I recall correctly)/RAM?

**Reply by thobitz** [reply]
Hi James, 

Thanks for the quick response. Regarding 6: These errors are follow-up errors from document generation errors earlier. I am building a repo which I synced from the Github master branch last week (still seems to be current), and that seems to have a few errors. After fixing them, the build runs through. Could you check whether the master branch builds for you? 

In general: Is it advisable to just download release ZIPs instead of cloning the repo?

Fixes: 

1) It looks like the variable DOT_GRAPH_MAX_NODES is set too low in the doxyfiles in OREData/doc and OREAnalytics/doc. But they are generated during build from OREAnalytics/doc/orea.doxy. After fixing that file (i.e. increasing DOT_GRAPH_MAX_NODES to 500), one error was gone.

2)   C:/sde/ore/ore/OREAnalytics/orea/aggregation/postprocess.hpp:62: warning: Found unknown command '\DELTA'
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MS
B8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\c8c5ed994f8075529695f9ad8f9e8f27\doc_orea.rule' exited with code -107
3741819. [C:\sde\ore\ore\build\OREAnalytics\doc\doc_orea.vcxproj]

Looks like doxygen stumbled over the \DELTA - maybe someone tried to write \Delta? I removed the comment and it built.

3)   C:/sde/ore/ore/QuantExt/qle/instruments/forwardbond.hpp:57: warning: Found unknown command '\Name'
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MS
B8066: Custom build for 'C:\sde\ore\ore\build\CMakeFiles\68fad14f002bf8ca171d655fdd364117\doc_quantext.rule' exited with code
-1073741819. [C:\sde\ore\ore\build\QuantExt\doc\doc_quantext.vcxproj]

Looks like a typo - \Name instead of \name. Fixed that, builds now. 

Best
Thomas

**Comment by thobitz** [comment]
Test Suite not running

Hi, 

The test suite (section 4.1: "ctest -C Release -j4 –timeout 3600") is not running:

1) Python: The following libraries are not mentioned explicitly and are required to run the tests:

pytest, datacompy, lxml, xmldiff, jsondiff

A requirements.txt for the tests (or installing them as part of the test script if not required elsewhere) might be a good idea.

2) 91 errors in the OREAnalytics tests because of expiry dates: 

C:\sde\ore\ore\OREAnalytics\test\cube.cpp(474): last checkpoint: "testSinglePrecisionJaggedCube" test entry
unknown location(0): fatal error: in "OREAnalyticsTestSuite/CubeTest/testDoublePrecisionJaggedCube": class QuantLib::Error: effective date (December 15th, 2016) later than or equal to termination date (December 15th, 2016)

Is there any "secret environment variable" or parameter I need to set for the effective date?

Best
Thomas

**Reply by francisduffy** [reply]
> 2\. 91 errors in the OREAnalytics tests because of expiry dates:

On this, I was noticing the same thing. I have opened a pull request for the test failures at https://github.com/OpenSourceRisk/Engine/pull/326. If you pull the changes in the branch for that pull request (or wait for it to be merged into the master branch) it should hopefully fix the OREAnalytics test quite errors.
